lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -243,6 +350,88 @@ void FIRRTLModuleLowering::runOnOperation() {
   circuit.erase();
 }
 
+void FIRRTLModuleLowering::lowerMemoryDecls(const SmallVector<FirMemory> &mems,
+                                            Block *top,
+                                            CircuitLoweringState &state) {
+  auto b =
+      ImplicitLocOpBuilder::atBlockBegin(UnknownLoc::get(&getContext()), top);
+  std::array<StringRef, 8> schemaFields = {
+      "depth",        "numReadPorts", "numWritePorts", "numReadWritePorts", "readLatency",
+      "writeLatency", "width",        "readUnderWrite"};
+  auto schemaFieldsAttr = b.getStrArrayAttr(schemaFields);
+  auto schema = b.create<rtl::RTLGeneratorSchemaOp>(
+      "FIRRTLMem", "FIRRTL_Memory", schemaFieldsAttr);
+  auto memorySchema = b.getSymbolRefAttr(schema);
+
+  Type b1Type = IntegerType::get(&getContext(), 1);
+
+  for (auto &mem : mems) {
+    SmallVector<rtl::ModulePortInfo> ports;
+    size_t inputPin = 0;
+    size_t outputPin = 0;
+
+    auto makePortCommon = [&](StringRef prefix, Twine i, Type bAddrType) {
+      ports.push_back({b.getStringAttr((prefix + "_clock_" + i).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr((prefix + "_en_" + i).str()), rtl::INPUT,
+                       b1Type, inputPin++});
+      ports.push_back({b.getStringAttr((prefix + "_addr_" + i).str()),
+                       rtl::INPUT, bAddrType, inputPin++});
+    };
+
+    Type bDataType;
+    if (mem.dataWidth > 0)
+      bDataType = IntegerType::get(&getContext(), mem.dataWidth);
+    Type bAddrType =
+        IntegerType::get(&getContext(), llvm::Log2_64_Ceil(mem.depth));
```

Comment:
`Log2_64_Ceil` is 0 when the depth is 1, which returns an `i0` type here. I think that edge case chokes #915. I haven't tracked down exactly where the mismatch is, but when using this PR and #915, I see this error:

```
'rtl.instance' op #2 operand type must be 'i0', but got 'i1'
...
note: see current operation: %mem0.ro_data_0 = "rtl.instance"(%clock, %arg2_valid, %0, %clock, %2, %1, %43, %arg0_data) {instanceName = "mem0", moduleName = @FIRRTLMem_1_1_0_32_1_0_1_1} : (i1, i1, i1, i1, i1, i1, i1, i32) -> i32
```

I know there is a PR out to canonicalize away 0-wide memories, but what's the approach for 1-wide memories? Should that be handled here? Should there be a separate pass to turn such memories into registers? Not sure what the thinking is, but there appears to be a mismatch between this PR and what #915 does.

Reply:
Confirmed this is just an edge case. When I updated my silly test to be a 2-wide memory instead of 1-wide, this pass and #915 work well together! Now to test it in the simulator...

Reply:
Works for me in the simulator.

Reply:
Nice.  I decided to make the minimum bus width always 1-bit (address, data), to keep the port signatures consistent.  Since FIRRTL already allows wider and narrower addresses than needed for a memory to be used, this seems well within the accepted behavior.

Reply:
Sounds good to me.